File: app/java/app/src/main/java/com/example/accounting/application/service/AuditLogService.java
package com.example.accounting.application.service;

import com.example.accounting.application.port.out.AuditLogRepository;
import com.example.accounting.domain.model.audit.AuditLog;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

/**
 * 監査ログサービス
 */
@Service
@Transactional(readOnly = true)
public class AuditLogService {

    private final AuditLogRepository auditLogRepository;

    public AuditLogService(AuditLogRepository auditLogRepository) {
        this.auditLogRepository = auditLogRepository;
    }

    /**
     * 監査ログを記録
     *
     * @param auditLog 監査ログ
     * @return 記録された監査ログ
     */
    @Transactional
    public AuditLog recordLog(AuditLog auditLog) {
        return auditLogRepository.save(auditLog);
    }

    /**
     * エンティティの監査ログを取得
     *
     * @param entityType エンティティ種別
     * @param entityId エンティティID
     * @return 監査ログリスト
     */
    public List<AuditLog> getLogsByEntity(String entityType, String entityId) {
        return auditLogRepository.findByEntity(entityType, entityId);
    }

    /**
     * ユーザーの監査ログを取得
     *
     * @param userId ユーザーID
     * @param startDate 開始日時
     * @param endDate 終了日時
     * @return 監査ログリスト
     */
    public List<AuditLog> getLogsByUser(String userId, LocalDateTime startDate, LocalDateTime endDate) {
        return auditLogRepository.findByUser(userId, startDate, endDate);
    }

    /**
     * 期間で監査ログを取得
     *
     * @param startDate 開始日時
     * @param endDate 終了日時
     * @param limit 取得件数上限
     * @return 監査ログリスト
     */
    public List<AuditLog> getLogsByPeriod(LocalDateTime startDate, LocalDateTime endDate, int limit) {
        return auditLogRepository.findByPeriod(startDate, endDate, limit);
    }

    /**
     * 全ての監査ログを取得
     *
     * @return 監査ログリスト
     */
    public List<AuditLog> getAllLogs() {
        return auditLogRepository.findAll();
    }
}

File: app/scala/src/main/scala/com/example/accounting/application/service/AuditLogService.scala
package com.example.accounting.application.service

import com.example.accounting.application.*
import com.example.accounting.application.port.in.AuditLogUseCase
import com.example.accounting.application.port.out.AuditLogRepository
import com.example.accounting.domain.audit.AuditLog
import scalikejdbc.*

import java.time.Instant

/**
 * 監査ログサービス（AuditLogUseCase の実装）
 */
class AuditLogService(
    auditLogRepository: AuditLogRepository,
) extends AuditLogUseCase:

  override def recordAuditLog(auditLog: AuditLog): Either[AppError, AuditLog] =
    try
      DB.localTx { implicit session =>
        val saved = auditLogRepository.save(auditLog)
        Right(saved)
      }
    catch
      case e: Exception =>
        Left(DatabaseError("監査ログの記録に失敗しました", Some(e)))

  override def getAuditLogsByEntity(entityType: String, entityId: String): Either[AppError, List[AuditLog]] =
    try
      DB.readOnly { implicit session =>
        val logs = auditLogRepository.findByEntity(entityType, entityId)
        Right(logs)
      }
    catch
      case e: Exception =>
        Left(DatabaseError("監査ログの取得に失敗しました", Some(e)))

  override def getAuditLogsByUser(
      userId: String,
      startDate: Instant,
      endDate: Instant,
  ): Either[AppError, List[AuditLog]] =
    try
      DB.readOnly { implicit session =>
        val logs = auditLogRepository.findByUser(userId, startDate, endDate)
        Right(logs)
      }
    catch
      case e: Exception =>
        Left(DatabaseError("監査ログの取得に失敗しました", Some(e)))

  override def getAuditLogsByPeriod(
      startDate: Instant,
      endDate: Instant,
      limit: Int,
  ): Either[AppError, List[AuditLog]] =
    try
      DB.readOnly { implicit session =>
        val logs = auditLogRepository.findByPeriod(startDate, endDate, limit)
        Right(logs)
      }
    catch
      case e: Exception =>
        Left(DatabaseError("監査ログの取得に失敗しました", Some(e)))

  override def getAuditLogsByAction(
      action: String,
      startDate: Instant,
      endDate: Instant,
  ): Either[AppError, List[AuditLog]] =
    try
      DB.readOnly { implicit session =>
        val logs = auditLogRepository.findByAction(action, startDate, endDate)
        Right(logs)
      }
    catch
      case e: Exception =>
        Left(DatabaseError("監査ログの取得に失敗しました", Some(e)))
